refactor(core): add `ApplicationRef.prototype.bootstrapImpl` with `injector` parameter

dgp1130 opened the pull request:
This allows any components individually bootstrapped to inherit from a unique `Injector`. This is useful when bootstrapping multiple root components with different providers.

Blocking [cl/741690539](http://cl/741690539).

AndrewKushnir commented on goldens/public-api/core/index.api.md:
```diff
@@ -134,9 +134,9 @@ export class ApplicationModule {
+    bootstrap<C>(component: Type<C>, rootSelectorOrNode?: string | any, injector?: Injector): ComponentRef<C>;
     // @deprecated
-    bootstrap<C>(componentFactory: ComponentFactory<C>, rootSelectorOrNode?: string | any): ComponentRef<C>;
+    bootstrap<C>(componentFactory: ComponentFactory<C>, rootSelectorOrNode?: string | any, injector?: Injector): ComponentRef<C>;
```
I think it'd be great if we add a new signature to mimic the contract of the `createComponent` API ([docs](https://angular.dev/api/core/createComponent)), i.e.:

```
bootstrap<C>(component: Type<C>, options: { environmentInjector: EnvironmentInjector; hostElement?: Element | undefined }): ComponentRef<C>;
```

The tricky part is that we'd need to determine what the second argument represents: a selector (a string), a DOM object or the "options" object, but we can probably do that by checking for a specific object shape to detect if it represents options.

dgp1130 replied:
Just to avoid unblocking internal use cases, I updated this to add a new, private `ɵbootstrap` function with the `injector` parameter.

I'll follow up on some potential designs to see if we can unify this with the existing `bootstrap` and `createComponent` so it becomes a part of our public API as well.

thePunderWoman commented:
This PR was merged into the repository by commit 7cb8639da9ca484ac823fc4a68d373ff25252971.

The changes were merged into the following branches: main, 19.2.x

dgp1130 commented:
@jessicajaniuk can you link me to the test failure? Not sure what would be going on here post-merge that would cause a failure.

thePunderWoman commented:
@dgp1130 I think you'll see it if you rebase first. Can you give that a try?

dgp1130 committed:
refactor(core): add `ApplicationRef.prototype.bootstrapImpl` with an `injector` parameter

This is a roll forward of commit d5a8a1c524805e047085ba02cf5d03f28b211fe1. Nothing is meaningfully different, as we're trying again to see if the CI failure is reproducible.

atscott commented:
This PR was merged into the repository by commit 2348f26d8a6d93770a9f1145c3354d399cf05057.

The changes were merged into the following branches: main, 19.2.x
